nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,159 @@ defmodule Nx.BinaryBackend do
+              padding_length = (m - k) * k
+
+              zero = number_to_binary(0, type)
+
+              h_bin <> String.duplicate(zero, padding_length)
+            else
+              dot_binary(q, shape_q, h_bin, shape_h, type)
+            end
+
+          r = dot_binary(h_bin, shape_q, r, shape_r, type)
+
+          {q, r}
+      end
+
+    q = from_binary(q_holder, q_bin)
+    r = from_binary(r_holder, r_bin)
+
+    case mode do
+      :complete ->
+        {q, r}
+
+      :reduced ->
+        {q, r[[0..(k - 1), 0..(n - 1)]]}
+    end
+  end
+
+  defp dot_binary(b1, shape1, b2, shape2, {_, s} = type) do
+    # matrix multiplication for when b1 and b2 are matrices
+    # represented in binary form
+    fun = fn lhs, rhs, acc ->
+      res = binary_to_number(lhs, type) * binary_to_number(rhs, type) + acc
+      {res, res}
+    end
+
+    v1 = aggregate_axes(b1, [1], shape1, s)
+    v2 = aggregate_axes(b2, [0], shape2, s)
+
+    for b1 <- v1, b2 <- v2, into: <<>> do
+      {bin, _acc} = bin_zip_reduce_axis(b1, b2, s, s, <<>>, 0, fun)
+      scalar_to_binary(bin, type)
+    end
+  end
+
+  defp householder_reflector(a_bin, num_rows_a, {_, num_bits} = type, target_k) do
+    # We receive the target shape so the binary is already pre-generated with 1s in the
+    # main diagonal for rows < k - n
+    <<head::bitstring-size(num_bits), tail::bitstring>> = a_bin
+
+    norm_a_squared =
+      for <<x_bin::bitstring-size(num_bits) <- a_bin>>, reduce: 0 do
+        acc ->
+          x = binary_to_number(x_bin, type)
+          acc + x * x
+      end
+
+    a_0 = binary_to_number(head, type)
+
+    v_0 =
+      if a_0 > 0 do
+        a_0 + :math.sqrt(norm_a_squared)
+      else
+        a_0 - :math.sqrt(norm_a_squared)
+      end
+
+    prefix_threshold = target_k - num_rows_a
+    prefix_size = prefix_threshold * num_bits
+    v_bin = <<0::size(prefix_size)>> <> number_to_binary(v_0, type) <> tail
+
+    # dot(v, v) = norm_v_squared, which can be calculated from norm_a as: norm_v_squared = norm_a_squared - a_0^2 + v_0^2
+
+    norm_v_squared = norm_a_squared - a_0 * a_0 + v_0 * v_0
+
+    scale = 2 / norm_v_squared
+
+    # execute I - 2 / norm_v_squared * outer(v, v)
+    {_, _, reflector_binary} =
+      for <<col_factor::bitstring-size(num_bits) <- v_bin>>,
+          <<row_factor::bitstring-size(num_bits) <- v_bin>>,
+          reduce: {0, 0, ""} do
+        {row, col, acc} ->
+          # The current element in outer(v, v) is given by col_factor * row_factor
+          # and the current I element is 1 when row == col
+          identity_element = if row == col, do: 1, else: 0
+
+          result =
+            if row >= prefix_threshold and col >= prefix_threshold do
+              identity_element -
+                scale * binary_to_number(col_factor, type) * binary_to_number(row_factor, type)
+            else
+              identity_element
+            end
+
+          acc = acc <> scalar_to_binary(result, type)
```

Comment:
`<>` copies both sides, so if it is often batter to build cons lists:

```suggestion
          acc = [acc | scalar_to_binary(result, type)]
```

then at the end: `IO.iodata_to_binary(reflector_binary)`.